org.eclipse.dartboard/src/org/eclipse/dartboard/launch/LaunchShortcut.java

```diff
@@ -38,57 +44,80 @@
  * @author Jonas Hungershausen
  *
  */
+@SuppressWarnings("restriction")
 public class LaunchShortcut implements ILaunchShortcut {
 
 	@Override
 	public void launch(ISelection selection, String mode) {
+		// Get the active project from the
 		IProject selected = null;
 		if (selection instanceof StructuredSelection) {
 			Object firstElement = ((StructuredSelection) selection).getFirstElement();
-			if (firstElement instanceof IProject) {
-				selected = (IProject) firstElement;
+			if (firstElement instanceof IResource) {
+				selected = ((IResource) firstElement).getProject();
 			}
 		}
 
+		// Shouldn't happen
+		if (selected == null) {
+			MessageDialog.openError(null, Messages.Launch_ConfigurationRequired_Title,
+					Messages.Launch_ConfigurationRequired_Body);
+			return;
+		}
+
+		launchProject(selected, mode);
+	}
+
+	@Override
+	public void launch(IEditorPart editor, String mode) {
+		IResource resource = editor.getEditorInput().getAdapter(IResource.class);
+
+		if (resource == null) {
+			MessageDialog.openError(null, Messages.Launch_ConfigurationRequired_Title,
+					Messages.Launch_ConfigurationRequired_Body);
+			return;
+		}
+
+		IProject project = resource.getProject();
+
+		launchProject(project, mode);
+	}
+
+	private void launchProject(IProject project, String mode) {
 		ILaunchManager manager = DebugPlugin.getDefault().getLaunchManager();
 
 		ILaunchConfigurationType type = manager.getLaunchConfigurationType(Constants.LAUNCH_CONFIGURATION_ID);
 
 		Set<String> modes = new HashSet<>();
 		modes.add(mode);
 		try {
+			// Find first launch configuration for selected project.
 			ILaunchConfiguration launchConfiguration = null;
 			for (ILaunchConfiguration conf : manager.getLaunchConfigurations(type)) {
-				if (conf.getAttribute(Constants.LAUNCH_SELECTED_PROJECT, "").equalsIgnoreCase(selected.getName())) { //$NON-NLS-1$
+				if (conf.getAttribute(Constants.LAUNCH_SELECTED_PROJECT, "").equalsIgnoreCase(project.getName())) { //$NON-NLS-1$
 					launchConfiguration = conf;
 				}
 			}
 
-			if (launchConfiguration != null) {
-				launchConfiguration.launch(mode, null);
-				return;
+			if (launchConfiguration == null) {
+				ILaunchConfigurationWorkingCopy copy = type.newInstance(null, manager.generateLaunchConfigurationName(
+						LaunchConfigurationsMessages.CreateLaunchConfigurationAction_New_configuration_2));
+
+				int result = DebugUIPlugin.openLaunchConfigurationEditDialog(PlatformUIUtil.getActiveShell(), copy,
```

Comment:
DebugUIPlugin is an internal class. We should use org.eclipse.debug.ui.DebugUITools.

Reply:
Done.